README.md
# liteServer
Very Lightweight Data Object Server. The basic principle is similar to the 
pvAccess protocol of EPICS. It hosts Lite Data Objects (**LDO**, analog of process variables in EPICS) and provides info/set/get/read/subscribe remote access to them using UDP protocol. 
Data encoding is implemented using UBJSON specification, which makes it very fast and efficient.

User responsibility is to provide private methods for connecting process variables to physical devices.

### Motivation
- Provide control for devices connected to non-linux machines. 
- The simplicity of the protocol makes it possible to implement in CPU-less FPGA device.
- The server is running on a remote machine. Device parameters can be 
manipulated using liteAccess.py.
- The liteAccess.py is the base class for accessing the server parameters.

### Bridged usage
To monitor and control liteServer-served devices from an existing architecture one can use or build a simple bridge:
##### A bridge for RHIC ADO Control architecture is liteServerMan.py
    liteServerMan.py -HmyHost myADO 
An ADO manager liteServerMan.py connects to a liteServer, running on myHost and creates the myADO. 
  - all input objects of the liteServer are translated to myADO input parameters
  - all output parameters of the myADO are translated to the liteServer objects
##### For EPICS Control architecture
The bridge liteServer-EPICS can be developed using a python-based implementation of IOC, for example:
[caproto](https://nsls-ii.github.io/caproto/)

### Installation
Prerequisites:
- **python3**: preferrably python3.6+ (it keeps dictionaries sorted)
- **py-ubjson**: versions 0.12+ provides greatly improved performance for multi-dimensional arrays
- **pyQT4 or pyQT5**: for GUI

### Features
 - Simplicity. The network protocol is **UDP**, error correction of late/missing/mangled data is
implemented. The serialization protocol is **UBJSON**: binary, easier than RPC, provides all JSON features. All this makes it possible to implement liteServer on a CPU-less FPGA.
 - Low latency, connection-less.
 - Supported requests:
   - **info()**, returns dictionary with information on requested LDOs and parameters
   - **get()**, returns dictionary of values of requested LDOs and parameters
   - **read()**, returns dictionary of all readable (non-constant) parameters of requested LDO
   - **set()**, set values or attributes of requested LDO parameters
   - **subscribe()** (not commissioned yet)
 - Multidimensional data (numpy arrays) are supported.
 - Access control info (username, program name) supplied in every request
 - Name service
   - file-based
   - network-based using a dedicated liteServer  (not commissioned yet)
 - Basic spreadsheet-based GUI: **ldoPet.py**
 - Architectures. All programs are 100% python. Tested on Linux and Windows.
 - Supported applications:
   - [Image analysis](https://github.com/ASukhanov/Imagin)

### Status
Commissioning of the second release.

## Examples
### liteScaler.py
Test server, providing two multi-channel scalers and a multi-dimesional array.
  To start the server:
  
    liteScaler.py
  To show and control liteScaler parameters using pre-configured spreadsheet liteScaler.yaml:

    ldoPet.py -f liteScaler.yaml
  


changelog
#``````````````````liteAccess`````````````````````````````````````````````````
#``````````````````Python2/3 compatibility````````````````````````````````````
from __future__ import print_function
from __future__ import unicode_literals
import sys
Python2 = sys.version_info.major == 2
if not Python2:
    basestring = str
    unicode = str
#,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
#__version__ = 'v01 2018-12-17'# created
#__version__ = 'v02 2018-12-17'# python3 compatible
#__version__ = 'v03 2018-12-19'#
#__version__ = 'v04 2018-12-26'# release
#__version__ = 'v05 2018-12-31'# timeout interception for sock.recvfrom
#__version__ = 'v06 2019-01-04'# socketSize increased
#__version__ = 'v07 2019-01-04'# -t timeout argument
#__version__ = 'v08 2019-01-06'# more detailed printing on timeout
#__version__ = 'v09 2019-01-17'# socket size set to UDP max (64k), timeout 0.1,
#__version__ = 'v10 2019-02-04'# bug fixed in main
#__version__ = 'v11 2019-05-21'# abridged printing
#__version__ = 'v12 2019-06-07'# TCP OK, debugging OK
#__version__ = 'v13 2019-06-07'# get(), set() info()
#__version__ = 'v14 2019-06-09'# numpy array support
#__version__ = 'v15 2019-06-10'# socket timeout defaulted to None. Be carefull with this setting 
#__version__ = 'v16 2019-06-10'# UDP Acknowledge
#__version__ = 'v17 2019-06-11'# chunking OK
#__version__ = 'v18 2019-06-17'# release, generic access to multiple or single items
#__version__ = 'v19 2019-06-28'# Dbg behavior fixed 
#__version__ = 'v20 2019-09-18'# try/except on pwd, avoid exception on default start
#__version__ = 'v21 2019-09-23'#
#__version__ = 'v22 2019-11-07'# --period
#__version__ = 'v23 2019-11-10'# reconnection
#__version__ = 'v24 2019-11-10'# Python2/3 compatible
#__version__ = 'v25 2019-11-10'# numpy array correctly decoded in multi-parameter requests
#__version__ = 'v26 2019-11-25'# get() returns {timestamp,value}, numpy decoding fixed 
#__version__ = 'v27a 2019-11-25'# firstValue()
#__version__ = 'v28 2019-11-27'# catch exception in execute_cmd
#__version__ = 'v29 2019-12-03'# firstValue python2/3 compatibility
#__version__ = 'v30 2019-12-06'# fix except: expectedArgWrong()
#__version__ = 'v31 2019-12-09'# socket timeout defaulted to 5
#__version__ = 'v32a 2019-12-10'# measurements() method for requesting only measurable parameters
#__version__ = 'v36 2020-02-06'# full re-design
#__version__ = 'v37 2020-02-09'# LdoPars info(), get(), read() set() are good.
#__version__ = 'v38 2020-02-10'# set() raising exceptions on failures
#__version__ = 'v39b 2020-02-11'# better error and timeout handling
#__version__ = 'v40 2020-02-13'# pid field added to request
#__version__ = 'v41 2020-02-14'# bug in set() fixed, deepcopy needed

#``````````````````liteServer`````````````````````````````````````````````````
#__version__ = 'v01 2018-12-14' # Created
#__version__ = 'v02 2018-12-15' # Simplified greatly using ubjson
#__version__ = 'v03 2018-12-17' # moved out the user-defined PVs, PV_all fixed
#__version__ = 'v04 2018-12-17' # python3
#__version__ = 'v05 2018-12-17'# more pythonic, using getattr,setattr
#__version__ = 'v06 2018-12-17'# get/set/info works, TODO: conserve type for set
#__version__ = 'v07 2018-12-21'# python2 compatibility, Device is method-less.
#__version__ = 'v08 2018-12-26'# python2 not supported, 
#__version__ = 'v09 2018-12-28'# PV.__init__ accepts parent, 
#Server.__init__ accepts host, port. No arguments in Server.loop().
#__version__ = 'v10 2019-01-03'# force parameters to be iterable, it simplifies the usage
#__version__ = 'v11 2019-01-17'# Device.__init__ checks if parameter is list.
#__version__ = 'v12 2019-05-23'# Raising extention, instead of printing. Special treatment of action parameters
#__version__ = 'v13 2019-05-23'# Tuple, tried use tuple for non writable, it does not work  
#__version__ = 'v14 2019-05-23'# opLimits
#__version__ = 'v15 2019-05-31'# count is array 
#__version__ = 'v16 2019-06-01'# Device 'server' incorporated
#__version__ = 'v17 2019-06-02'# DevDict is OrderedDict
#__version__ = 'v19 2019-06-09'# numpy array support
#__version__ = 'v20 2019-06-10'# UDP Acknowledge
#__version__ = 'v20 2019-06-10'# framing works 
#__version__ = 'v21 2019-06-10'# chunking OK
#__version__ = 'v22a 2019-06-17'# release
#__version__ = 'v23a 2019-06-21'# redesign
#__version__ = 'v24 2019-11-07'# release
#__version__ = 'v25 2019-11-08'# PV.parent removed, it conflicts with json
#__version__ = 'v26 2019-11-10'# Dbg and devDict are Server attributes

PVsheet.py
#!/usr/bin/env python3
"""Spreadsheet view of process variables from a remote liteServer."""

#__version__ = 'v00 2019-05-06' # it is just sceleton, pvview.pvv is not processed
#__version__ = 'v01 2019-05-06' # using pg.TableWidget()
#__version__ = 'v02 2019-05-20' # using QTableWidget, with checkboxes
#__version__ = 'v03 2019-05-21' # PVTable,
#__version__ = 'v04 2019-05-22' # bool PVs treated as checkboxes
#__version__ = 'v05 2019-05-29' # spinboxes in table
#__version__ = 'r06 2019-05-29' # first release
#__version__ = 'r07 2019-05-30' # cell spanning is OK
#__version__ = 'r08 2019-05-31' # release 08
#__version__ = 'v09 2019-05-31' # detection of right click
#__version__ = 'v10 2019-06-01' # pargs.file
#__version__ = 'v11 2019-06-02' # automatic generation of the pvsheet.tmp
#__version__ = 'v12 2019-06-02' # boolean action is OK: don't set checkbox to the same state 
#TODO 1) discrete parameters, set array
#__version__ = 'v13 2019-06-03' # is_spinbox check for writable, is_bool
#__version__ = 'v14 2019-06-07' # dbg corrected
#__version__ = 'v15 2019-06-07' # pv.values is dict
#__version__ = 'v16 2019-06-09'# numpy array support
#__version__ = 'v18 2019-06-09'# spinbox fixed
#__version__ = 'v19 2019-06-21'# redesign
#TODO: not very reliable on wifi (wide network)
#__version__ = 'v20 2019-06-21'#
#__version__ = 'v21 2019-06-27'# right click fixed, config file OK
#__version__ = 'v22 2019-06-28'# table created with initial values, not titles
#__version__ = 'v23 2019-09-15'#
__version__ = 'v24 2019-11-21'#

import threading, socket, subprocess, sys, time
from timeit import default_timer as timer
from collections import OrderedDict as OD
from PyQt5 import QtCore, QtGui, QtWidgets
import numpy as np
import traceback
import liteAccess as LA

EventExit = threading.Event()

#````````````````````````````Helper functions`````````````````````````````````
def printw(msg): print('WARNING: '+msg)
def printe(msg): print('ERROR: '+msg)
def printd(msg): 
    if pargs.dbg:
        print('DBG:'+str(msg))

def ip_address():
    """Platform-independent way to get local host IP address"""
    return [(s.connect(('8.8.8.8', 53)), s.getsockname()[0], s.close())\
        for s in [socket.socket(socket.AF_INET, socket.SOCK_DGRAM)]][0][1]
#,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
class QDoubleSpinBoxPV(QtWidgets.QDoubleSpinBox):
    """Spinbox, which stores associated PV""" 
    def __init__(self,pv):
        super().__init__()
        self.pv = pv
        try:    opl = self.pv.opLimits['values']
        except: pass
        else:
            self.setRange(*opl)
            ss = (opl[1]-opl[0])/100.
            #ss = round(ss,12)# trying to fix deficit 1e-14, not working
            self.setSingleStep(ss)
        self.valueChanged.connect(self.handle_value_changed)
        #print('instantiated %s'%self.pv.title())
        
    def handle_value_changed(self):
        #print('handle_value_changed')
        #print('changing %s to '%self.pv.title()+str(self.value()))
        try:
            self.pv.v = self.value()
        except Exception as e:
            print(e)
            
    def contextMenuEvent(self,event):
        # we don't need its contextMenu (activated on right click)
        print('RightClick at spinbox with PV %s'%self.pv.name)
        mainWidget.rightClick(self.pv)
        pass

class myTableWidget(QtWidgets.QTableWidget):
    def mousePressEvent(self,*args):
        button = args[0].button()
        item = self.itemAt(args[0].pos())
        try:
            row,col = item.row(),item.column()
        except:
            return
        if button == 2: # right button
            try:
                pv = pvTable.pos2obj[(row,col)]
                #print('RightClick at PV %s.'%pv.name)
                mainWidget.rightClick(pv)
            except:
                pass
        else:
            super().mousePressEvent(*args)

class Window(QtWidgets.QWidget):
    def __init__(self, rows, columns):
        QtWidgets.QWidget.__init__(self)
        self.table = myTableWidget(rows, columns, self)
        self.table.setShowGrid(False)
        for row in range(rows):
          spanStart,spanFilled = None, False
          self.table.setRowHeight(row,20)
          if pvTable.pos2obj[(row,0)] is None:
            continue
          colSkip = 0
          for col in range(columns):
            #print('row,col',(row,col))
            try: obj = pvTable.pos2obj[(row,col)]
            except Exception as e:
                printd('Not an object:'+str(e))
                continue
            colOut = col - colSkip

            # check if not a PV object
            #print('obj[%i,%i]:'%(row,col)+str(type(obj)))
            if not isinstance(obj,PV):
                if isinstance(obj,str):
                    if len(obj) == 0: continue
                    if obj[0] == '[':
                        spanStart = col
                        #print('span starts at [%i,%i]:%s'%(row,spanStart,obj))
                        continue
                    elif obj[0] == ']':
                        colSkip += 1
                        #print('span ends at [%i,%i]'%(row, col))
                        self.table.setSpan(row,spanStart,1,col-spanStart)
                        continue
                    if spanStart is not None and not spanFilled:
                        spanFilled = True
                        colOut = spanStart
                    item = QtWidgets.QTableWidgetItem(str(obj))
                    item.setForeground(QtGui.QBrush(QtGui.QColor('darkBlue')))
                    self.table.setItem(row, colOut, item)
                elif isinstance(obj,QPushButtonCmd):
                    printd('pushButton at [%i,%i]'%(row, colOut))
                    self.table.setCellWidget(row, colOut, obj)
                continue
            
            if spanStart is not None:
                spanStart = None
                
            # the object is PV
            pv = obj
            #val = pv.v
            printd('pv.initialValue of %s:'%pv.name+str(pv.initialValue))
            initialValue = pv.initialValue['value'][0]
            #print( 'initialValue',initialValue)
            pvTable.par2pos[pv] = row,colOut
            try:
                item = QtWidgets.QTableWidgetItem(pv.title())
            except Exception as e:
                printw('could not define Table[%i,%i]'%(row,colOut))
                print(str(e))
                print('Traceback: '+repr(traceback.format_exc()))
                continue
            #print('ok',item)
            #print('pvTable [%i,%i] is %s %s'%(row,colOut,pv.title(),type(pv)))
            try:
                if pv.guiType == 'bool':
                    #print( 'PV %s is boolean:'%pv.name+str(initialValue))
                    #item.setText(pv.name.split(':')[1])
                    item.setText(pv.name.split(':',1)[1])
                    item.setFlags(QtCore.Qt.ItemIsUserCheckable |
                                  QtCore.Qt.ItemIsEnabled)
                    state = QtCore.Qt.Checked if initialValue else QtCore.Qt.Unchecked
                    item.setCheckState(state)
                    self.table.setCellWidget(row, colOut, item)
                    continue
                    
                elif pv.guiType == 'spinbox':
                    printd('it is spinbox:'+pv.title())
                    spinbox = QDoubleSpinBoxPV(pv)
                    # using other ways is more complicated as it is not trivial
                    # to transfer argument to the method
                    #spinbox = QtWidgets.QDoubleSpinBox(self\
                    #,valueChanged=self.value_changed)
                    
                    spinbox.setValue(float(initialValue))
                    self.table.setCellWidget(row, colOut, spinbox)
                    continue
            except Exception as e:
                #printw('in is_bool '+pv.title()+':'+str(e))
                pass
            self.table.setItem(row, col, item)

        #self.table.itemClicked.connect(self.handleItemClicked)
        #self.table.itemPressed.connect(self.handleItemPressed)
        #self.table.itemDoubleClicked.connect(self.handleItemDoubleClicked)
        self.table.cellClicked.connect(self.handleCellClicked)
        #self.table.cellDoubleClicked.connect(self.handleCellDoubleClicked)
        
        layout = QtWidgets.QVBoxLayout(self)
        layout.addWidget(self.table)
        self._list = []
        monitor = PVMonitor()

    def closeEvent(self,*args):
        # Called when the window is closed
        print('>closeEvent')
        EventExit.set()

    def handleItemPressed(self, item):
        print('pressed[%i,%i]'%(item.row(),item.column()))

    def handleItemDoubleClicked(self, item):
        print('DoubleClicked[%i,%i]'%(item.row(),item.column()))

    def handleItemClicked(self, item):
        print('clicked[%i,%i]'%(item.row(),item.column()))
        self.handleCellClicked(item.row(),item.column())

    def handleCellDoubleClicked(self, x,y):
        print('cell DoubleClicked[%i,%i]'%(x,y))

    def handleCellClicked(self, row,column):
        item = self.table.item(row,column)
        printd('cell clicked[%i,%i]:'%(row,column))
        pv = pvTable.pos2obj[row,column]
        if isinstance(pv,str):
            return
        try:
            if pv.guiType =='bool':
                checked = item.checkState() == QtCore.Qt.Checked
                printd('bool clicked '+pv.name+':'+str(checked))
                pv.v = checked # change server's pv
            else:
                d = QtWidgets.QDialog(self)
                d.setWindowTitle("Info")
                pname = pv.title()
                ql = QtWidgets.QLabel(pname,d)
                qte = QtWidgets.QTextEdit(item.text(),d)
                qte.move(0,20)
                #d.setWindowModality(Qt.ApplicationModal)
                d.show()
        except Exception as e:
            printe('exception in handleCellClicked: '+str(e))

    def update(self,a):
        print('mainWidget update',a)
        tableItem = self.table.item(2,1)
        try:
            tableItem.setText(str(a[0]))
        except Exception as e:
            printw('in tableItem.setText:'+str(e))
            
    def rightClick(self,pv):
        #print('mainWidget. RightClick on %s'%pv.name)
        d = QtWidgets.QDialog(self)
        pname = pv.title()
        d.setWindowTitle("Info on PV %s"%pname)
        attributes = pv.attributes()
        #print('attributes:%s'%str(attributes)[:100])
        txt = '    Attributes:\n'
        for attr,v in attributes.items():
            #vv = v if isinstance(v,str) else list(v)[0]
            vv = v if isinstance(v,str) else list(v)
            if vv is None:
                continue
            if isinstance(vv,list):
                vv = vv[:100]
            txt += attr+':\t'+str(vv)+'\n'
        qte = QtWidgets.QLabel(txt,d)
        qte.setWordWrap(True)
        d.resize(300,150)
        d.show()

#`````````````````````````````````````````````````````````````````````````````
myslotBusy = False
def MySlot(a):
    """Global redirector of the SignalSourceDataReady"""
    global myslotBusy
    printd('MySlot received event:'+str(a))
    if myslotBusy:
        print('Busy')
        return
    myslotBusy = True
    if mainWidget is None:
        printe('mainWidget not defined yet')
        return
    for pv,rowCol in pvTable.par2pos.items():
        printd('updating PV '+pv.name)
        if isinstance(pv,str):
            printw('logic error')
            continue
        try:
            val = pv.v['value']
            printd('val:%s'%str(val)[:100])
            if isinstance(val,list):
                if pv.guiType =='bool':
                    #print('PV '+pv.name+' is bool = '+str(val))
                    state = window.table.item(*rowCol).checkState()
                    if val[0] != (state != 0):
                        printd('flip')
                        window.table.item(*rowCol).setCheckState(val[0])
                    continue
                elif pv.guiType == 'spinbox':
                    continue
                # standard PV
                #print('PV '+pv.name+' is list[%i] of '%len(val)\
                #+str(type(val[0])))
                txt = str(val)[1:-1] #avoid brackets
            elif isinstance(val,str):
                printd('PV '+pv.name+' is text')
                txt = val
                #continue
            elif isinstance(val,np.ndarray):
                txt = '%s: %s'%(val.shape,str(val))
            else:
                txt = 'Unknown type of '+pv.name+'='+str(type(val))
                printw(txt+':'+str(val))
            window.table.item(*rowCol).setText(txt)
        except Exception as e:
            printw('updating [%i,%i]:'%rowCol+str(e))
            print('Traceback: '+repr(traceback.format_exc()))
    myslotBusy = False
#,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
#````````````````````````````Data provider
class PVMonitor(QtCore.QThread): 
    # inheritance from QtCore.QThread is needed for qt signals
    SignalSourceDataReady = QtCore.pyqtSignal(object)
    def __init__(self):
        # for signal/slot paradigm we need to call the parent init
        super(PVMonitor,self).__init__()
        #...
        thread = threading.Thread(target=self.thread_proc)
        thread.start()
        self.SignalSourceDataReady.connect(MySlot)

    def thread_proc(self):
        printd('>thread_proc')
        while not EventExit.isSet():
            self.callback(None)
            EventExit.wait(2)
        print('<thread_proc')

    def callback(self,*args):
        #print('cb:',args)
        self.SignalSourceDataReady.emit(args)
#,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
class PV():
    """Process Variable object, provides getter and setter"""
    def __init__(self,name):
        self.name = name
        printd('pv name: '+str(name))
        self.pv = LA.PV(name.split(':'))
        self.key = list(self.pv.info())[0]
        printd('key:'+str(self.key))
        #print('pv vars:'+str(vars(self.pv)))
        self.initialValue = self.v # use getter to store initial value
        self.t = 0.
        # creating attributes from remote ones
        self.attr = self.pv.info()[self.key]
        print('attrs %s'%self.attr)
        #for attribute,v in self.attr.items():
        #    if attribute not in ['count', 'features', 'opLimits']:                
        #        continue
        #    print('Creating attribute %s.%s = '%(name,attribute)+str(v))
        #    setattr(self,attribute,v)
        self.guiType = self.gui_type()
        print('type of %s:'%self.name+str(self.guiType))
            
    @property
    def v(self): # getter
        # return values and store timestamp
        #v = self.pv.value[self.key]['value']
        v = self.pv.value[self.key]
        printd('getter of %s called, value type:'%self.name+str(type(v)))
        return(v)

    @v.setter
    def v(self, value):
        printd('setter of %s'%self.name+' called to change PV to '+str(value))
        self.pv.value = value

    @v.deleter
    def v(self):
        print("deleter of v called")
        del self.initialValue

    def title(self): return self.name

    def gui_type(self):
        iv = self.initialValue['value']
        print('iv',self.name,str(iv)[:60])
        if len(iv) != 1:
            return None
            
        if isinstance(iv[0],bool):
            return 'bool'

        if self.is_writable():
            if type(iv[0]) in (float,int):
                return 'spinbox'                
            if 'legalValues' in self.attr:
                return 'combobox'
        return None
        
    def is_writable(self):
        return 'W' in self.attr['features']
            
    def attributes(self):
        return self.attr

class QPushButtonCmd(QtWidgets.QPushButton):
    def __init__(self,text,cmd):
        self.cmd = cmd
        super().__init__(text)
        self.clicked.connect(self.handleClicked)
        
    def handleClicked(self):
        #print('clicked',self.cmd)
        p = subprocess.Popen(self.cmd, stdout=subprocess.PIPE, shell=True)

class PVTable():
    """PV table maps: parameter to (row,col) and (row,col) to object"""
    def __init__(self,fileName):
        
        self.par2pos = OD()
        self.pos2obj = OD()
        maxcol = 0
        if pargs.pvfile is None:
            pargs.pvfile = self.build_temporary_pvfile()
        with open(pargs.pvfile,'r') as infile:
            row = 0
            for line in infile:
                #line = line[:-1]# remove eol
                line = line.rstrip()
                if len(line) == 0:  continue
                if line[0] == '#':  continue
                if line[0] == '!':
                    self.pos2obj[(row,0)] = None
                    row += 1  
                    continue
                printd('%3i:'%row+line)
                cols = line.split(',')
                nCols = len(cols)
                for col,token in enumerate(cols):
                  if True:#try:
                    #print( 'token:'+str(token))
                    if len(token) == 0:
                        obj = ''
                    elif token in '[]':
                        obj = token
                    elif token[0] in ('"',"'"):
                        blank,txt,attributeString = token.split(token[0],2)
                        if len(attributeString) == 0:
                            printd('cell is text')
                            obj = txt
                        else: # the cell is text with attributes
                            #print( 'cell is text with attributes') 
                            action,cmd = attributeString.split(':',1)
                            action = action[1:]
                            if action == 'launch':
                                #print('pushButton created with cmd:%s'%cmd)
                                obj = QPushButtonCmd(txt,cmd)
                    elif '`' in token: # PV's attribute
                        raise NotImplementedError('syntax quotation ` is not supported yet in '+token)
                    #    printd('check for attribute '+token)
                    #    pvname,attrib = token.split('`')
                    #    pv = PV(pvname,access)
                    #    printd('temporary PV %s created'%pvname)
                    #    obj = str(getattr(pv,attrib))
                    #    if obj[0] == '[': obj = obj[1:]
                    #    if obj[-1] == ']': obj = obj[:-1]
                    else: # the cell is PV
                        #print( 'the "%s" is pv'%token)
                        if True:#try:
                            obj = PV(token)
                        else:#except Exception as e:
                            txt = 'Cannot create PV %s:'%token+str(e)
                            raise NameError(txt)
                    self.pos2obj[(row,col)] = obj
                    #print(row,col,type(obj))
                  else:#except Exception as e:
                    printw(str(e))
                    self.pos2obj[(row,col)] = '?'

                maxcol = max(maxcol,nCols)
                row += 1
        self.shape = row,maxcol
        print('table created, shape: '+str(self.shape))

    def print_PV_at(self,row,col):
        try:
            pv = self.pos2obj[row,col]
            v = pv.v
            txt = str(v) if len(v) <= 10 else str(v[:10])[:-1]+',...]'
            print('Table[%i,%i]:'%(row,col)+pv.name+' = '+txt)
        except Exception as e:
            printw('in print_PV:'+str(e))

    def print_loc_of_PV(self,pv):
        try:
            row,col = self.par2pos[pv]
            print('Parameter '+pv.name+' is located at Table[%i,%i]:'%(row,col))
        except Exception as e:
            printw('in print_loc:'+str(e))

    def build_temporary_pvfile(self):
        fname = 'pvsheet.tmp'
        print('>build_temporary_pvfile')
        pvServerInfo = LA.PV([pargs.host]).info()
        deviceSet = {i.split(':')[0] for i in pvServerInfo.keys()}
        printd('devs:'+str(deviceSet))
        f = open(fname,'w')
        #for dev in deviceSet:
        curDev = ''
        for devParName,pardict in pvServerInfo.items():
            dev,parName = devParName.split(':')
            if dev != curDev:
                curDev = dev
                f.write("[,'____Device: %s____',]\n"%dev)
            f.write("'%s',%s\n"%(parName,devParName))
        f.close()
        print('PV spreadsheet config file generated: %s'%fname)
        return fname
#`````````````````````````````````````````````````````````````````````````````
if __name__ == '__main__':
    global mainWidget
    mainWidget = None
    import sys
 
    import argparse
    parser = argparse.ArgumentParser(description = __doc__)
    parser.add_argument('-d','--dbg', action='store_true', help='debugging')
    parser.add_argument('-p','--port',default='9999',
      help='Port number')
    parser.add_argument('-H','--host',default=ip_address(),nargs='?',
      help='Hostname')
    parser.add_argument('-t','--timeout',type=float,default=None,
      help='timeout of the receiving socket')
    parser.add_argument('pvfile', nargs='?', 
      help='PV list description file')
    pargs = parser.parse_args()
    printd('Monitoring of PVs at '+pargs.host+':%s'%pargs.port)

    app = QtWidgets.QApplication(sys.argv)

    # read config file
    pvTable = PVTable(pargs.pvfile)

    # define GUI
    window = Window(*pvTable.shape)
    try:
        title = 'PVs from '+socket.gethostbyaddr(pargs.host)[0].split('.')[0]
    except:
        title = 'PVs from '+pargs.host
    #print(title)
    window.setWindowTitle(title)
    window.resize(350, 300)
    window.show()
    mainWidget = window

    # arrange keyboard interrupt to kill the program
    import signal
    signal.signal(signal.SIGINT, signal.SIG_DFL)
    
    # start GUI
    try:
        app.instance().exec_()
        #sys.exit(app.exec_())
    except KeyboardInterrupt:
        # This exception never happens
        print('keyboard interrupt: exiting')
        EventExit.set()
    print('Application exit')

